File: code/R/src/rl_agent/q_learning.R
# An Reinforcement Learning Application for Statistical Arbitrage
# karol.przybylak.13@ucl.ac.uk, kmprzybylak@gmail.com

# Qlearning

#moze agent moze uzywac t-1 i t, zamiast t i t+1



# -----------------------------------------------------------------
#region Dependencies



#endregion 1
# -----------------------------------------------------------------
#region Function Definition

Qcontrol <- function(pair, coef, features, Nepisodes, cost, pretrained_agent = NULL, ...){
  # Initialize all elements of the RL framework
  list[cut_points, actions, alpha , epsilon, N_0, episode, Ns, Nsa, Qsa,
       Action_track] <- InitializeRLframework(features, 'qlearning')
  # Initialize additional parameters

  params <- list(...)
  env    <- environment()
  check_for_opt_args(params, env)
  if(!exists("verbose", envir = env))  verbose <- FALSE
  if(!exists("rwrd_fun", envir = env)) rwrd_fun <- 'dsr'
  if(verbose) ns_vars(env, values = T)
    # ENDED HERE



  # If available, continue with the policy from the training period
  if(!is.null(pretrained_agent)) list[Qsa, Action_track, Nsa, Ns, epsilon, alpha, cut_points] <- pretrained_agent

  while(episode <= Nepisodes){
    i <- 1
    while (i < nrow(pair)){

      state <- StateIndexes(features[i,], cut_points) # get the current state
      action <- ChooseAction(Qsa, state, epsilon, actions) # choose action from the current state
      action_dim <- match(action, actions)
      state_action <- as.numeric(c(state, action_dim))

      Action_track <- c(Action_track, action)
      counts <- UpdateN(state, Ns, Nsa, actions, action, cut_points)
      Ns <- counts[[1]]
      Nsa <- counts[[2]]

      # Update constants
      # EPsilon should also depend on the value of the present state, could grow bigger if negative state values
      # But maybe they have been evaluated to such a value? But maybe it should grow with growing volatility or if
      # a regime switch was detected. Or otherwise: sudden changes in values indicate changing conditions and should
      # create a bigger epsilon again.
      # epsilon <- N_0 / (N_0 + Ns[matrix(state, 1)])
      epsilon <- 1/(log(Nsa[matrix(state_action, 1)])+1)
      #alpha <- 1/Nsa[matrix(state_action, 1)]
      alpha <- 1/(log(Nsa[matrix(state_action, 1)])+1)

      state_next          <- StateIndexes(features[i+1,], cut_points) # get the following state
      action_target       <- ChooseAction(Qsa, state_next, 0, actions) # deterministicaly choose the best action
      action_target_dim   <- match(action_target, actions)
      state_action_target <- as.numeric(c(state_next, action_target_dim))

      #To choose raw returns instead of Diff Sharpe Ratio it's needed to remove explicit reward_fun argument
      # For DSR reward function eta_dsr argument has to be set, otherwise the default value, 0.5, will be imposed


      rwrd <- Qreward(pair[c(i:(i + 1)), ], coef, action, cost, rwrd_fun = rwrd_fun, eta_dsr = 0.2)

      # Qlearning backup
      Qsa[matrix(state_action, 1)] <- Qsa[matrix(state_action, 1)] + alpha * (rwrd + Qsa[matrix(state_action_target, 1)] - Qsa[matrix(state_action, 1)])

      i <- i + 1
    }
   episode <- episode + 1
  }

  return(list(Qsa = Qsa, Action_track = Action_track, Nsa = Nsa, Ns = Ns,
              epsilon = epsilon, alpha = alpha, cut_points = cut_points))
}

File: code/python/src/rl_agent_old/q_learning.py
"""
Q-learning implementation for statistical arbitrage.
"""

import numpy as np

from src.rl_agent_python.reward import q_reward
from src.rl_agent_python.utils import (
    choose_action,
    initialize_rl_framework,
    state_indexes,
    update_n,
)


class QLearning:
    def __init__(self, features: np.ndarray, actions: list[int] = [-1, 0, 1]):
        """
        Initialize Q-learning agent.

        Args:
            features: Feature matrix for state representation
            actions: List of possible actions (default: [-1, 0, 1] for short, neutral, long)
        """
        self.actions = actions
        self.features = features

        # Initialize RL framework
        (
            self.cut_points,
            self.actions,
            self.alpha,
            self.epsilon,
            self.n_0,
            self.episode,
            self.ns,
            self.nsa,
            self.qsa,
            self.action_track,
        ) = initialize_rl_framework(features, "qlearning")

    def train(
        self,
        pair: np.ndarray,
        coef: np.ndarray,
        n_episodes: int,
        cost: float,
        pretrained_agent: dict | None = None,
        verbose: bool = False,
        reward_func: str = "dsr",
        **kwargs,
    ) -> dict:
        """
        Train the Q-learning agent.

        Args:
            pair: Price data for the trading pair
            coef: Cointegration coefficients
            n_episodes: Number of training episodes
            cost: Transaction cost
            pretrained_agent: Optional pretrained agent to continue training from
            verbose: Whether to print training progress
            reward_func: Reward function to use ('dsr' or 'raw_return')
            **kwargs: Additional parameters

        Returns:
            Dictionary containing trained agent parameters
        """
        # Load pretrained agent if provided
        if pretrained_agent is not None:
            self.qsa = pretrained_agent["qsa"]
            self.action_track = pretrained_agent["action_track"]
            self.nsa = pretrained_agent["nsa"]
            self.ns = pretrained_agent["ns"]
            self.epsilon = pretrained_agent["epsilon"]
            self.alpha = pretrained_agent["alpha"]
            self.cut_points = pretrained_agent["cut_points"]

        # Training loop
        while self.episode <= n_episodes:
            i = 0
            while i < len(pair) - 1:
                # Get current state
                state = state_indexes(self.features[i], self.cut_points)

                # Choose action
                action = choose_action(self.qsa, state, self.epsilon, self.actions)
                action_dim = self.actions.index(action)
                state_action = np.array([*state, action_dim])

                # Track action
                self.action_track.append(action)

                # Update visit counts
                counts = update_n(
                    state, self.ns, self.nsa, self.actions, action, self.cut_points
                )
                self.ns, self.nsa = counts

                # Update epsilon and alpha
                self.epsilon = 1 / (np.log(self.nsa[tuple(state_action)] + 1))
                self.alpha = 1 / (np.log(self.nsa[tuple(state_action)] + 1))

                # Get next state and best action
                state_next = state_indexes(self.features[i + 1], self.cut_points)
                action_target = choose_action(self.qsa, state_next, 0, self.actions)
                action_target_dim = self.actions.index(action_target)
                state_action_target = np.array([*state_next, action_target_dim])

                # Calculate reward
                reward = q_reward(
                    pair[i : i + 2],
                    coef,
                    action,
                    cost,
                    reward_func=reward_func,
                    eta_dsr=0.2,
                )

                # Q-learning update
                self.qsa[tuple(state_action)] += self.alpha * (
                    reward
                    + self.qsa[tuple(state_action_target)]
                    - self.qsa[tuple(state_action)]
                )

                i += 1

            self.episode += 1

        return {
            "qsa": self.qsa,
            "action_track": self.action_track,
            "nsa": self.nsa,
            "ns": self.ns,
            "epsilon": self.epsilon,
            "alpha": self.alpha,
            "cut_points": self.cut_points,
        }
